(window.webpackJsonp=window.webpackJsonp||[]).push([[10],{355:function(t,e,a){"use strict";a.r(e);var s=a(42),o=Object(s.a)({},(function(){var t=this,e=t.$createElement,a=t._self._c||e;return a("ContentSlotsDistributor",{attrs:{"slot-key":t.$parent.slotKey}},[a("h1",{attrs:{id:"using-magpurify2"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#using-magpurify2"}},[t._v("#")]),t._v(" Using MAGpurify2")]),t._v(" "),a("h2",{attrs:{id:"quick-start"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#quick-start"}},[t._v("#")]),t._v(" Quick start")]),t._v(" "),a("div",{staticClass:"language-bash extra-class"},[a("pre",{pre:!0,attrs:{class:"language-bash"}},[a("code",[t._v("magpurify2 composition test_data/genomes/* output\nmagpurify2 coverage test_data/genomes/* output --bam_files test_data/bam_files/*\nmagpurify2 taxonomy test_data/genomes/* output magpurify2DB\nmagpurify2 filter test_data/genomes/* output filtered_genomes\n")])])]),a("h2",{attrs:{id:"the-composition-module"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#the-composition-module"}},[t._v("#")]),t._v(" The "),a("code",[t._v("composition")]),t._v(" module")]),t._v(" "),a("p",[a("code",[t._v("magpurify2 composition")]),t._v(" is the command used to identify putative contaminants using tetranucleotide frequencies.")]),t._v(" "),a("div",{staticClass:"language- extra-class"},[a("pre",{pre:!0,attrs:{class:"language-text"}},[a("code",[t._v("usage: magpurify2 composition [-h] [-s STRICTNESS] [-t THREADS] [--quiet] genomes [genomes ...] output_directory\n\npositional arguments:\n  genomes               Input genomes in the FASTA format.\n  output_directory      Directory to write the output files to.\n\noptional arguments:\n  -h, --help            show this help message and exit\n  -s STRICTNESS, --strictness STRICTNESS\n                        Strictness of the contaminant detection algorithm. Must be a number between 0 (less strict) and 1 (more strict). (default: 0.5)\n  -t THREADS, --threads THREADS\n                        Number of threads to use. All by default. (default: 4)\n  --quiet               Suppress the logger output (default: False)\n")])])]),a("h2",{attrs:{id:"the-coverage-module"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#the-coverage-module"}},[t._v("#")]),t._v(" The "),a("code",[t._v("coverage")]),t._v(" module")]),t._v(" "),a("div",{staticClass:"custom-block tip"},[a("p",{staticClass:"custom-block-title"},[t._v("How to generate the BAM files")]),t._v(" "),a("p",[t._v("BAM files store reads that map to the target metagenome or MAG and are processed by MAGpurify2 to estimate the coverage of each contig. To generate the BAM required by MAGpurify2 you should first map your reads to the complete metagenome using a proper tool (such as "),a("a",{attrs:{href:"https://github.com/BenLangmead/bowtie2",target:"_blank",rel:"noopener noreferrer"}},[t._v("Bowtie 2"),a("OutboundLink")],1),t._v(", "),a("a",{attrs:{href:"https://github.com/lh3/minimap2",target:"_blank",rel:"noopener noreferrer"}},[t._v("minimap2"),a("OutboundLink")],1),t._v(" or "),a("a",{attrs:{href:"https://github.com/bwa-mem2/bwa-mem2",target:"_blank",rel:"noopener noreferrer"}},[t._v("BWA-MEM2"),a("OutboundLink")],1),t._v(") and then convert and sort the output using "),a("a",{attrs:{href:"https://github.com/samtools/samtools",target:"_blank",rel:"noopener noreferrer"}},[t._v("samtools"),a("OutboundLink")],1),t._v(".")]),t._v(" "),a("div",{staticClass:"language-bash extra-class"},[a("pre",{pre:!0,attrs:{class:"language-bash"}},[a("code",[t._v("$ "),a("span",{pre:!0,attrs:{class:"token function"}},[t._v("mkdir")]),t._v(" bt2\n$ bowtie2-build --threads "),a("span",{pre:!0,attrs:{class:"token number"}},[t._v("4")]),t._v(" metagenome.fna bt2/metagenome.fna\n$ bowtie2 --threads "),a("span",{pre:!0,attrs:{class:"token number"}},[t._v("4")]),t._v(" -x bt2/metagenome.fna "),a("span",{pre:!0,attrs:{class:"token punctuation"}},[t._v("\\")]),t._v("\n  -1 sample1_R1.fastq.gz -2 sample1_R2.fastq.gz "),a("span",{pre:!0,attrs:{class:"token punctuation"}},[t._v("\\")]),t._v("\n  "),a("span",{pre:!0,attrs:{class:"token operator"}},[t._v("|")]),t._v(" samtools "),a("span",{pre:!0,attrs:{class:"token function"}},[t._v("sort")]),t._v(" -@ "),a("span",{pre:!0,attrs:{class:"token number"}},[t._v("4")]),t._v(" -o sample1.bam -\n")])])]),a("p",[t._v("We reccomend mapping the reads to the metagenome and not directly to the MAGs and this is because of two factors:")]),t._v(" "),a("ul",[a("li",[t._v("When you map the reads into the MAG a read that was originated from the sequencing of a closely related genome might be erroneously aligned to the MAG (cross-mapping), introducing bias to the coverage estimation.")]),t._v(" "),a("li",[t._v("Metagenome-wide mappings can be used to estimate the coverage of all the contigs in the metagenome, thus allowing MAGpurify2 to process multiple MAGs in a single execution.")])])])])}),[],!1,null,null,null);e.default=o.exports}}]);