(window.webpackJsonp=window.webpackJsonp||[]).push([[12],{378:function(e,n,t){"use strict";t.r(n);var s=t(17),i=Object(s.a)({},(function(){var e=this,n=e.$createElement,t=e._self._c||n;return t("ContentSlotsDistributor",{attrs:{"slot-key":e.$parent.slotKey}},[t("h1",{attrs:{id:"introduction"}},[t("a",{staticClass:"header-anchor",attrs:{href:"#introduction"}},[e._v("#")]),e._v(" Introduction")]),e._v(" "),t("p",[t("strong",[e._v("ADD ILLUSTRATION: genomes ➔ contigs ➔ binning ➔ magpurify2")])]),e._v(" "),t("p",[e._v("Metagenome-assembled genomes (MAGs) are draft genome sequences that are recovered from metagenomes containing genomic fragments of multiple species. The retrieval of MAG sequences from metagenomes is accomplished through the clustering — in this context, also known as binning — of contigs that share similar sequence composition and exhibit correlated sequencing coverage across different environmental samples. Due to technical limitations inherent to sequencing, metagenomic assembly and binning, MAGs are usually fragmented, incomplete, and have varying degrees of contamination (I.e. sequences from a different organism). Under the typical assumption that a MAG represents a composite of a single species, contaminant sequences are especially concerning as they will cause unreliable biological interpretations.")]),e._v(" "),t("p",[e._v("MAGpurify2 aims to improve the quality of MAGs through the identification and removal of contaminant sequences, improving the reliability of downstream analysis of the genomic sequences. It can be be used to process thousands of genomes at once, making it especially useful for large-scale studies.")])])}),[],!1,null,null,null);n.default=i.exports}}]);