(window.webpackJsonp=window.webpackJsonp||[]).push([[9],{364:function(e,n,i){"use strict";i.r(n);var t=i(17),s=Object(t.a)({},(function(){var e=this,n=e.$createElement,i=e._self._c||n;return i("ContentSlotsDistributor",{attrs:{"slot-key":e.$parent.slotKey}},[i("h1",{attrs:{id:"introduction"}},[i("a",{staticClass:"header-anchor",attrs:{href:"#introduction"}},[e._v("#")]),e._v(" Introduction")]),e._v(" "),i("p",[i("strong",[e._v("ADD ILLUSTRATION: genomes ➔ contigs ➔ binning ➔ magpurify2")])]),e._v(" "),i("p",[e._v("Metagenome-assembled genomes (MAGs) are genomic sequences extracted from broader metagenomic assemblies. This is acomplished through the binning of contigs that share similar sequence composition and exhibit correlated sequencing coverage across different environmental conditions.")]),e._v(" "),i("p",[e._v("Due to technical limitations of sequencing, metagenomic assembly and binning, MAGs are usually fragmented, incomplete and display varying degrees of contamination (i.e. sequences from a different organism). Under the typical assumption that a given MAG represents a single species, this last issue is especially concearning as contaminant sequences will lead to unreliable biological interpretations.")]),e._v(" "),i("p",[e._v("MAGpurify2 aims to improve the quality of MAGs through the identification and removal of contaminant sequences, improving the reliability of downstream analysis of the genomic sequences.")])])}),[],!1,null,null,null);n.default=s.exports}}]);