<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Using MAGpurify2 | MAGpurify2</title>
    <meta name="generator" content="VuePress 1.5.4">
    
    <meta name="description" content="">
    <link rel="preload" href="/magpurify2/assets/css/0.styles.e2e1409a.css" as="style"><link rel="preload" href="/magpurify2/assets/js/app.a0eb5e6b.js" as="script"><link rel="preload" href="/magpurify2/assets/js/2.64cecbbb.js" as="script"><link rel="preload" href="/magpurify2/assets/js/12.9a5ebe1f.js" as="script"><link rel="prefetch" href="/magpurify2/assets/js/10.8d826114.js"><link rel="prefetch" href="/magpurify2/assets/js/11.48fac3ec.js"><link rel="prefetch" href="/magpurify2/assets/js/3.7d566c34.js"><link rel="prefetch" href="/magpurify2/assets/js/4.d74c9bdb.js"><link rel="prefetch" href="/magpurify2/assets/js/5.712311aa.js"><link rel="prefetch" href="/magpurify2/assets/js/6.7e9bfd8a.js"><link rel="prefetch" href="/magpurify2/assets/js/7.3ebea663.js"><link rel="prefetch" href="/magpurify2/assets/js/8.b20cc351.js"><link rel="prefetch" href="/magpurify2/assets/js/9.7eab64cd.js">
    <link rel="stylesheet" href="/magpurify2/assets/css/0.styles.e2e1409a.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/magpurify2/" class="home-link router-link-active"><!----> <span class="site-name">MAGpurify2</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <nav class="nav-links can-hide"><div class="nav-item"><a href="https://github.com/apcamargo/magpurify2" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Repository
  <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <!----></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="https://github.com/apcamargo/magpurify2" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Repository
  <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></div> <!----></nav>  <ul class="sidebar-links"><li><a href="/magpurify2/docs/" aria-current="page" class="sidebar-link">Introduction</a></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading"><span>Theory</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/magpurify2/docs/codon-usage-theory.html" class="sidebar-link">Codon usage module</a></li><li><a href="/magpurify2/docs/composition-theory.html" class="sidebar-link">Composition module</a></li><li><a href="/magpurify2/docs/coverage-theory.html" class="sidebar-link">Coverage module</a></li><li><a href="/magpurify2/docs/contaminant-detection.html" class="sidebar-link">Contaminant detection</a></li></ul></section></li><li><a href="/magpurify2/docs/installation.html" class="sidebar-link">Installation</a></li><li><a href="/magpurify2/docs/usage.html" aria-current="page" class="active sidebar-link">Using MAGpurify2</a></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="using-magpurify2"><a href="#using-magpurify2" class="header-anchor">#</a> Using MAGpurify2</h1> <h2 id="generating-bins"><a href="#generating-bins" class="header-anchor">#</a> Generating bins</h2> <p><em>Recommend binners and DAS Tool.</em></p> <h2 id="quick-start"><a href="#quick-start" class="header-anchor">#</a> Quick start</h2> <div class="language-bash extra-class"><pre class="language-bash"><code>magpurify2 composition test_data/genomes/* output
magpurify2 coverage test_data/genomes/* output --bam_files test_data/bam_files/*
magpurify2 taxonomy test_data/genomes/* output magpurify2DB
magpurify2 filter test_data/genomes/* output filtered_genomes
</code></pre></div><h2 id="the-composition-module"><a href="#the-composition-module" class="header-anchor">#</a> The <code>composition</code> module</h2> <p><code>magpurify2 composition</code> is the command used to identify putative contaminants using tetranucleotide frequencies.</p> <div class="language- extra-class"><pre class="language-text"><code>usage: magpurify2 composition [-h] [-s STRICTNESS] [-t THREADS] [--quiet] genomes [genomes ...] output_directory

positional arguments:
  genomes               Input genomes in the FASTA format.
  output_directory      Directory to write the output files to.

optional arguments:
  -h, --help            show this help message and exit
  -s STRICTNESS, --strictness STRICTNESS
                        Strictness of the contaminant detection algorithm. Must be a number between 0 (less strict) and 1 (more strict). (default: 0.5)
  -t THREADS, --threads THREADS
                        Number of threads to use. All by default. (default: 4)
  --quiet               Suppress the logger output (default: False)
</code></pre></div><h2 id="the-coverage-module"><a href="#the-coverage-module" class="header-anchor">#</a> The <code>coverage</code> module</h2> <div class="custom-block tip"><p class="custom-block-title">How to generate the BAM files</p> <p>BAM files store reads that map to the target metagenome or MAG and are processed by MAGpurify2 to estimate the coverage of each contig. To generate the BAM required by MAGpurify2 you should first map your reads to the complete metagenome using a proper tool (such as <a href="https://github.com/BenLangmead/bowtie2" target="_blank" rel="noopener noreferrer">Bowtie 2<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>, <a href="https://github.com/lh3/minimap2" target="_blank" rel="noopener noreferrer">minimap2<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a> or <a href="https://github.com/bwa-mem2/bwa-mem2" target="_blank" rel="noopener noreferrer">BWA-MEM2<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>) and then convert and sort the output using <a href="https://github.com/samtools/samtools" target="_blank" rel="noopener noreferrer">samtools<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>.</p> <div class="language-bash extra-class"><pre class="language-bash"><code>$ <span class="token function">mkdir</span> bt2
$ bowtie2-build --threads <span class="token number">4</span> metagenome.fna bt2/metagenome
$ bowtie2 --threads <span class="token number">4</span> -x bt2/metagenome <span class="token punctuation">\</span>
  -1 sample1_R1.fastq.gz -2 sample1_R2.fastq.gz <span class="token punctuation">\</span>
  <span class="token operator">|</span> samtools <span class="token function">sort</span> -@ <span class="token number">4</span> -o sample1.bam -
</code></pre></div><p>We recommend mapping the reads to the metagenome (superset) and not directly to the MAGs (subsets). This is due to two factors:</p> <ul><li>When you map the reads into the MAG a read that was originated from the sequencing of a closely related genome might be erroneously aligned to the MAG (cross-mapping), introducing bias to the coverage estimation.</li> <li>Metagenome-wide mappings can be used to estimate the coverage of all the contigs in the metagenome, thus allowing MAGpurify2 to process multiple MAGs in a single execution.</li></ul> <p>If the target MAGs are derived from multiple source metagenomes you need input BAM files containing read mappings to each one of them.</p></div> <p>If you don't have access to the raw sequencing data or to previously generated BAM files you can input coverage data stored in a tab-separated values (TSV) file. To do so, you should use the <code>--coverage_file</code> argument:</p> <div class="language- extra-class"><pre class="language-text"><code>coverage genomes/* output --coverage_file contig_coverages.tsv --threads 4
</code></pre></div><p>The first column of the coverage file must store the contigs names. The remaining columns should contain the coverage of each contig across multiple samples, as shown in the example below:</p> <div class="language- extra-class"><pre class="language-text"><code>contig_1     15.744    12.605    25.148    3.728    0.000
contig_2     34.466    48.019    18.222    3.707    4.195
contig_3      0.000    22.356    21.944    4.479    4.463
contig_4     14.201     9.993     0.000    0.925    4.608
contig_5     17.179    12.280    56.643    3.586    4.226
contig_6      5.239     8.430    5.2070    3.988    0.000
contig_7     17.737    16.005    29.692    4.039    4.190
contig_8      0.000    15.866    13.663    0.877    2.269
contig_9     19.129    15.145    21.249    0.000    2.342
contig_10    11.074     9.574    19.673    0.934    4.732
</code></pre></div></div> <footer class="page-edit"><!----> <!----></footer> <div class="page-nav"><p class="inner"><span class="prev">
      ←
      <a href="/magpurify2/docs/installation.html" class="prev">
        Installation
      </a></span> <!----></p></div> </main></div><div class="global-ui"><div class="vuepress-eqn"></div><span class="vuepress-eq"></span></div></div>
    <script src="/magpurify2/assets/js/app.a0eb5e6b.js" defer></script><script src="/magpurify2/assets/js/2.64cecbbb.js" defer></script><script src="/magpurify2/assets/js/12.9a5ebe1f.js" defer></script>
  </body>
</html>
